src/checkers/inference/model/serialization/ToStringSerializer.java

```diff
@@ -266,6 +267,41 @@ public String serialize(ArithmeticConstraint arithmeticConstraint) {
         return sb.toString();
     }
 
+    @Override
+    public String serialize(ImplicationConstraint implicationConstraint) {
+        boolean prevShowVerboseVars = showVerboseVars;
+        showVerboseVars = false;
+        // format(non-empty): assumption1 & assumption 2 & ... & assumption n -> conclusion
+        // format(empty): [ ] -> conclusion
+        final StringBuilder sb = new StringBuilder();
+        sb.append(getCurrentIndentString());
+        int oldIndentationLevel = indentationLevel;
+        // This is to avoid indentation for each sub constraint that comprises the implicationConstraint
+        indentationLevel = 0;
+        String assumptions = getAssumptionsString(implicationConstraint);
+        String conclusion = implicationConstraint.getConclusion().serialize(this);
+        sb.append(assumptions).append(" -> ").append(conclusion);
+        // Recover the previous indentation level to not affect further serializations
+        indentationLevel = oldIndentationLevel;
+        showVerboseVars = prevShowVerboseVars;
+        return sb.toString();
+    }
+
+    private String getAssumptionsString(ImplicationConstraint implicationConstraint) {
+        int length = implicationConstraint.getAssumptions().size();
+        if (length == 0) {
+            return "[ ]";
+        }
+
+        StringBuilder sbAssumptions = new StringBuilder();
+        for (int i = 0; i < length - 1; i++) {
```

Comment:
would it be cleaner to use [String.join()](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.Iterable-) and avoid this unusual loop stop condition?

Reply:
Done